### PR TITLE
Requestor API routes

### PR DESCRIPTION
**Pull request by stephlow**
Same changes as #86, but somehow the git history got borked there so I just copied over the changes to a new branch to get it out of the way

### CODE REVIEW

File: fpx/src/api/client.rs [hunk 1]
```diff
@@ -48,14 +48,19 @@ impl ApiClient {
         &self,
         method: Method,
         path: impl AsRef<str>,
+        body: Option<String>,
```
Comment (hatchan) [inline fpx/src/api/client.rs]: This `body` should be a `body: Option<B>` and then we can add `B: serde::ser::Serialize`.

File: fpx/src/api/client.rs [hunk 2]
```diff
@@ -48,14 +48,19 @@ impl ApiClient {
...
         T: serde::de::DeserializeOwned,
         E: serde::de::DeserializeOwned,
     {
         let u = self.base_url.join(path.as_ref())?;
 
-        let req = self.client.request(method, u);
+        let req = self
+            .client
+            .request(method, u)
+            .header("Content-Type", "application/json")
```
Comment (hatchan) [inline fpx/src/api/client.rs]: We should only set this if `body` is some.

File: fpx/src/api/client.rs [hunk 3]
```diff
@@ -88,7 +93,28 @@ impl ApiClient {
...
+    pub async fn request_list(
+        &self,
+    ) -> Result<Vec<models::RequestSummary>, ApiClientError<RequestListError>> {
+        let path = "api/requests";
+
+        self.do_req(Method::GET, path, None).await
+    }
+
+    /// Create and execute a new request
+    pub async fn request_post(
```
Comment (hatchan) [inline fpx/src/api/client.rs]: `request_create` is a better name

File: fpx/src/api/client.rs [hunk 4]
```diff
@@ -88,7 +93,28 @@ impl ApiClient {
...
+        &self,
+        new_request: models::NewRequest,
+    ) -> Result<models::Response, ApiClientError<models::NewRequestError>> {
+        let path = "/api/requests";
+
+        let json = serde_json::to_string(&new_request).unwrap();
```
Comment (hatchan) [inline fpx/src/api/client.rs]: This part can be moved in the generic `do_req` fn

File: fpx/src/api/handlers/requests.rs [hunk 5]
```diff
@@ -1,23 +1,143 @@
+use std::collections::BTreeMap;
+use std::str::FromStr;
+use std::time::Duration;
+
```
Comment (hatchan) [inline fpx/src/api/handlers/requests.rs]: We can merge these import blocks

File: fpx/src/commands/client/requests.rs [hunk 6]
```diff
@@ -42,3 +48,78 @@ async fn get_request(args: GetArgs) -> Result<()> {
...
+}
+
+/// Temporary inline this code into this module
+///
+#[derive(clap::Args, Debug)]
+pub struct PostArgs {
+    /// Base url of the fpx dev server.
+    #[arg(from_global)]
+    pub base_url: Url,
+    /// HTTP method to use (e.g., GET, POST, PUT, DELETE).
```
Comment (hatchan) [inline fpx/src/commands/client/requests.rs:75]: Can you add some newlines between de property and the documentation for some better readability

File: fpx/src/data.rs [hunk 7]
```diff
@@ -228,37 +227,90 @@ impl Store {
...
             .await?;
 
-        Ok(request.id)
+        Ok(request)
+    }
+
+    pub async fn request_list<T: ReadTransaction>(
+        &self,
+        tx: &Transaction<'_, T>,
+    ) -> Result<Vec<models::Request>> {
```
Comment (hatchan) [inline fpx/src/data.rs:251]: We should make the model more public, probably can just add `pub` to the struct.

File: fpx/src/api/client.rs [hunk 8]
```diff
@@ -113,6 +140,6 @@ impl ApiClient {
     ) -> Result<Vec<models::Span>, ApiClientError<SpanGetError>> {
         let path = format!("api/traces/{}/spans", trace_id.as_ref());
 
-        self.do_req(Method::GET, path).await
+        self.do_req(Method::GET, path, None::<()>).await
```
Comment (hatchan) [inline fpx/src/api/client.rs:166]: Not a big fan of having to define these `::<()>` things, but this is fine for now. We can solve this another time.

File: fpx/src/api/handlers/requests.rs [hunk 9]
```diff
@@ -1,24 +1,142 @@
-use crate::api::errors::{ApiServerError, CommonError};
-use crate::api::models::Request;
+use std::{collections::BTreeMap, str::FromStr, time::Duration};
```
Comment (hatchan) [inline fpx/src/api/handlers/requests.rs]: This can be split and merged with the rest

File: fpx/src/commands/client/requests.rs [hunk 10]
```diff
@@ -12,13 +12,21 @@ pub struct Args {
...
 pub enum Command {
-    /// Get a requests
+    /// Get a request
     Get(GetArgs),
+
+    /// List requests
+    List(ListArgs),
+
+    /// Create and execute a request
+    Post(PostArgs),
```
Comment (hatchan) [inline fpx/src/commands/client/requests.rs]: Think create would be a better name for this. And giving that we are also executing it, maybe `call` would be better?

File: fpx/src/commands/client/requests.rs [hunk 11]
```diff
@@ -42,3 +50,82 @@ async fn get_request(args: GetArgs) -> Result<()> {
 
     Ok(())
 }
+
+/// Temporary inline this code into this module
```
Comment (hatchan) [inline fpx/src/commands/client/requests.rs]: This comment can be removed (and others like it). Initially I was under the impression that we should create separate files for each get/post/invoke. But I think it is fine to have 1 file for each of the commands per resource.

**Review by hatchan — APPROVED** [review]
Some minor things, but overall looks good :+1:

File: fpx/src/api/client.rs (hunk 8 above)
Comment (stephlow) [reply fpx/src/api/client.rs:166]: Wasn't a fan either, but couldn't figure out a way to make it compile without these type annotations :(

**Pull request merged**